File: src/edit/features/Marker/lib/marker/Marker.js
import SplitToken from './SplitToken'

export class Marker {
	constructor() {
		this.tokenFactories = []
		this.treeMiddlewares = []

		this.join = true
		this.blankFactory = null
	}

	async blank({text}) {
		if (!this.blankFactory) {
			throw new Error(`Blank factory is not defined`)
		}
		const token = await this.blankFactory.create(text)
		return token.compile()
	}

	addTokenFactory(tokenFactory, {blank = false} = {}) {
		this.tokenFactories.push(tokenFactory)
		if (blank) {
			this.blankFactory = tokenFactory
		}
		return this
	}

	addTreeMiddleware(callback) {
		this.treeMiddlewares.push(callback)
		return this
	}

	async evaluate(text) {
		let tokens = text.split('\n')
		tokens = this.markupTokens(tokens)
		if (this.join) {
			tokens = this.joinTokens(tokens)
		}

		return compileTokens(tokens)
		// tree = lxml.html.fromstring(compiled)
		// return await self.process_tree(tree)
	}

	joinTokens(tokens) {
		return mergeTokens(joinTokens(tokens))
	}

	async processTree(tree) {
		// for fn in self.treeMiddlewares:
		// tree = await fn(tree)
		// return tree
	}

	markupTokens(tokens) {
		const markedTokens = []

		tokens.forEach(data => {
			const token = this.createToken(data)
			if (!token) {
				throw new Error(`Token not recognized for data ${data}`)
			}

			markedTokens.push(token)
		})

		return markedTokens
	}

	createToken(data) {
		for (let factory of this.tokenFactories) {
			if (factory.test(data)) {
				return factory.create(data)
			}
		}

		return null
	}
}

function joinTokens(tokens) {
	const isSplit = x => x instanceof SplitToken

	let bufferTag = null
	let tokensBuffer = []
	let tokensList = []
	for (let token of tokens) {
		const tag = token.name
		const changed = bufferTag !== null && tag !== bufferTag

		let flush = isSplit(token)
		if (changed) {
			flush = true
		}

		if (!token.isJoinable()) {
			flush = true
		}

		if (flush) {
			if (tokensBuffer.length) {
				tokensList.push(tokensBuffer)
			}

			if (isSplit(token)) {
				tokensBuffer = []
				bufferTag = null
			} else {
				tokensBuffer = [token]
				bufferTag = tag
			}
		} else {
			bufferTag = tag
			tokensBuffer.push(token)
		}
	}

	if (tokensBuffer.length) {
		tokensList.push(tokensBuffer)
	}

	return tokensList
}

function mergeTokens(tokensList) {
	return tokensList
		.map(ts => ts
			.reduce(
				(token, x) => token.merge(x)
			)
		)
}

async function compileTokens(tokens) {
	return Promise.all(
		tokens.map(x => x.compile())
	)
}

File: modules/sync/kazimir/Marker.py
import lxml.html
from markdown import markdown

from kazimir.Token import SplitToken


class Marker:
    def __init__(self) -> None:
        super().__init__()
        self.tokens = []

    def add_token_factory(self, token):
        self.tokens.append(token)

    def parse(self, text: str):
        tokens = text.split('\n')
        tokens = self.markup_tokens(tokens)
        tokens = join_tokens(tokens)
        tokens = merge_tokens(tokens)

        return tokens

    def markup_tokens(self, tokens: list):
        marked_tokens = []

        for data in tokens:
            t = self.create_token(data)
            if t:
                marked_tokens.append(t)
            else:
                raise Exception(f'Token not recognized for data {data}')
        return marked_tokens

    def create_token(self, data):
        for f in self.tokens:
            if f.token.test(data):
                return f.create(data=data)
        return None


def join_tokens(tokens):
    is_split = lambda x: isinstance(x, SplitToken)

    buffer_tag = None
    buffer = []
    ts = []
    for token in tokens:
        tag = token.name
        changed = buffer_tag is not None and tag != buffer_tag

        flush = is_split(token)
        if changed:
            flush = True

        if not token.joinable:
            flush = True

        if flush:
            if buffer:
                ts.append(buffer)
            if is_split(token):
                buffer = []
                buffer_tag = None
            else:
                buffer = [token]
                buffer_tag = tag
        else:
            buffer_tag = tag
            buffer.append(token)

    if len(buffer):
        ts.append(buffer)
    return ts


def merge_tokens(tokens: list):
    tokens_out = []
    for ts in tokens:
        token = ts[0]
        for x in ts[1:]:
            token = token.merge(x)
        tokens_out.append(token)
    return tokens_out
